Invoice and payment void reasons

irinatodeva opened the pull request:
This converts the reasons used for invoice and payment void methods from text string to a classified FS::reason
The changes are more related to the FS object's method interfaces than the freeside UI. The current UI used to void payments and invoices is untouched but there are changes in the navigation (page redirects) to stay on the payment_history page when payment apply / unapply / void / unvoid was invoked. Also a new section in the menu was added to manage the void reasons in Configuration --> Billing.

freeside commented:
Thanks for the pull request.  I agree that anything that's currently a text reason can be moved to the new reason system.

Please include a complete pull request including UI changes.  If we're going to do this, invoice and payment void should act that other things that have reasons and allow you to select or autocreate a reason.  See httemplate/elements/tr-select-reason.html

Should we consider having separate invoice, credit and payment void reasons instead of one type for all of them?  I'm not sure one way or the other.

freeside commented:
Also, existing text reasons should be migrated on upgrade.  FS::reason_Mixin has an _upgrade_reasonnum subroutine you can use.  Take a look at the _upgrade_data subroutines in FS::cust_credit and FS::cust_refund for examples.

irinatodeva commented:
The different reason types ('Void payment' vs 'Void credit' vs 'Void invoice') are optional and will be used only if a reason is passed like a legacy code would do (by the reason text) and not with the FS::reason object as an argument.

I have seen the _upgrade_reasonnum and used it in places where there was already an _upgrade_data method in the classes. I'll make changes to make sure that this upgrade reason will be properly set for payments and invoices (as well as package line items).

irinatodeva commented:
Hi,

I added all the changes you requested. It's worth mentioning that this is renaming a permission check. Previously the tr-select-reason.html would add the option to create a new reason based on access_righ.rightname 'Add on-the-fly void credit reason' which is now renamed to the more accurate 'Add on-the-fly void reason' since it's used in the void credit, void payment and void invoice iframes.
I wrote an email earlier today to Mark Wells and you were cc-ed with some questions but I ended up commiting the code the in the pull request and this way you could review them and discuss directly what the code is doing and why.

Regards,
Irina Todeva

freeside commented:
Irina,

Thank you for submitting this. I've merged it to our development branch.

Your fix for the _upgrade_reasonnum method seems to be correct. The "#31702" was an internal ticket number; the issue we were dealing with was the reason() method being used to set the reasonnum field, which I was in the process of removing, but I missed the relevance of it to the _upgrade_reasonnum procedure. Good catch. After merging that fix, I was able to migrate the invoice and payment void reasons on my test instance.

Best regards,

-- 
Mark Wells
Senior Developer
Freeside Internet Services, Inc.
